### PR TITLE
DB: task pre-held status + retry and poll timers

### LINKED ISSUE #620: shutdown without setting all waiting tasks to the held state
This slows run-db shutdown in large suites, see:

https://github.com/cylc/cylc/issues/616#issuecomment-22154720
https://github.com/cylc/cylc/issues/616#issuecomment-22165112

Current normal shutdown procedure is: set all waiting tasks to held to prevent any more tasks from being submitted, then when all tasks are held or succeeded (i.e. none submitted or running) shut the suite down.

It should be just as easy, or easier!, to do this: don't submit any more tasks, and when no tasks are submitted or running shut the suite down.

### LINKED ISSUE #2050: Tasks which are held lose their held status on a restart
In the process of debugging suite tasks it can be very useful to hold a suite while an individual task is rerun multiple times.

Sometimes the task can be set up so that it succeeds, but not actually run as required. For instance, changing a forecast task so that it ends on the timestep before it fails means that diagnostic fields can be written out for in depth analysis in order to track down numerical instabilities or bugs. In this instance it is vital to hold the suite and run only the required tasks.

This process is painfully slow, especially if the failure occurs in a large job that takes a lot of compute resource. This sort of task can very easily take multiple weeks, so the suite will likely timeout over the weekend, and at the moment when a suite is restarted on a Monday morning it will immediately try and run all of the previously held tasks.

Because of this, tasks which are held need to remain held when the suite is restarted.

### PR DESCRIPTION
**Pull request by matthewrmshin**
New persistence across restart:
* pre-held status of held tasks.
* execution and submission retry states and timers of tasks.
* execution and submission poll timers of tasks.

Don't hold tasks that are beyond stop point (and don't release them to main pool).

Close #620. Close #2050.

@hjoliver @arjclark please review.

### CODE REVIEW

File: lib/cylc/rundb.py [hunk 1]
```diff
@@ -709,6 +714,16 @@ def take_checkpoints(self, event, other_daos=None):
                     dao.tables[table_name + "_checkpoints"].add_insert_item(
                         [id_] + list(row))
 
+    def upgrade_from_611(self):
```
Comment (arjclark) [inline lib/cylc/rundb.py]: Shouldn't the upgrade routine be handling the change from `TABLE_TASK_EVENT_HANDLER_TRY_STATES` to `TABLE_TASK_ACTION_TIMERS` as well?

Comment (arjclark) [inline lib/cylc/rundb.py]: Ah, well remembered!

File: lib/cylc/task_pool.py [hunk 2]
```diff
@@ -237,35 +236,23 @@ def add_to_runahead_pool(self, itask):
             self.log.warning(
                 itask.identity +
                 ' cannot be added to pool: task ID already exists')
-            del itask
```
Comment (matthewrmshin) [inline lib/cylc/task_pool.py]: They don't do anything! The `del` statement would remove the symbol `itask` from the local scope (and hence reduces the reference count by 1 to the data it references), but given that:

1. `itask` is passed into the method as an argument, the actual data is likely to be still referenced by the caller, so it is not like we can really free up the memory here.
2. Both statements are followed by a `return` statement, so the symbol will be out of scope any way.

**Comment by matthewrmshin** [comment]
Oh well. Looks like it is necessary to store the global suite held flag after all.

**Comment by hjoliver** [comment]
@matthewrmshin - I thought you'd asked a question about that, but had seemingly deleted it (or was it just somewhere else?)?   The distinction between "suite held" and "all tasks held" as per my previous comment seems a useful one to me - it's not uncommon to want to trigger a single task in a held suite without its successor potentially going off too.

**Comment by matthewrmshin** [comment]
Hi @hjoliver I deleted that from the main comment in #2050, naively thinking that it wasn't necessary. I'll see what I can do this week.

**Comment by hjoliver** [comment]
I guess that could be a follow-up enhancement if you like.  This is a good improvement even without it.

**Comment by matthewrmshin** [comment]
Hi @hjoliver I have pushed a small change that should fix the behaviour you reported. (Manually tested as working in my environment.) I'll try and add a test to the battery if I get a chance today.

**Pull request merged**